File: src/types/subscription.ts
export interface SubscriptionPlan {
  id: number;
  name: string;
  description?: string;
  price: number;
  billing_interval: 'monthly' | 'yearly';
  features: string[];
  max_students?: number;
  max_teachers?: number;
  max_courses?: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Subscription {
  id: number;
  admin_id: number;
  plan_id: number;
  status: 'active' | 'cancelled' | 'expired';
  start_date: string;
  end_date: string;
  auto_renew: boolean;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionInvoice {
  id: number;
  subscription_id: number;
  amount: number;
  status: 'paid' | 'pending' | 'failed';
  billing_date: string;
  paid_date?: string;
  payment_method?: string;
  invoice_number: string;
  created_at: string;
}

export interface UsageMetrics {
  id: number;
  admin_id: number;
  metric_name: string;
  metric_value: number;
  recorded_at: string;
}

export interface CreateSubscriptionPlan {
  name: string;
  description?: string;
  price: number;
  billing_interval: 'monthly' | 'yearly';
  features: string[];
  max_students?: number;
  max_teachers?: number;
  max_courses?: number;
  is_active?: boolean;
}

export interface UpdateSubscriptionPlan {
  name?: string;
  description?: string;
  price?: number;
  billing_interval?: 'monthly' | 'yearly';
  features?: string[];
  max_students?: number;
  max_teachers?: number;
  max_courses?: number;
  is_active?: boolean;
}

export interface CreateSubscription {
  admin_id: number;
  plan_id: number;
  status: 'active' | 'cancelled' | 'expired';
  start_date: string;
  end_date: string;
  auto_renew?: boolean;
}

export interface UpdateSubscription {
  plan_id?: number;
  status?: 'active' | 'cancelled' | 'expired';
  end_date?: string;
  auto_renew?: boolean;
}

export interface CreateSubscriptionInvoice {
  subscription_id: number;
  amount: number;
  status: 'paid' | 'pending' | 'failed';
  billing_date: string;
  paid_date?: string;
  payment_method?: string;
  invoice_number: string;
}

export interface UpdateSubscriptionInvoice {
  status?: 'paid' | 'pending' | 'failed';
  paid_date?: string;
  payment_method?: string;
}

export interface CreateUsageMetrics {
  admin_id: number;
  metric_name: string;
  metric_value: number;
}

File: app/schemas/subscription.py
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel

class SubscriptionPlanBase(BaseModel):
    name: str
    description: Optional[str] = None
    price: float
    billing_interval: str  # monthly, yearly
    features: List[str]
    max_students: Optional[int] = None
    max_teachers: Optional[int] = None
    max_courses: Optional[int] = None
    is_active: bool = True

class SubscriptionPlanCreate(SubscriptionPlanBase):
    pass

class SubscriptionPlanUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    billing_interval: Optional[str] = None
    features: Optional[List[str]] = None
    max_students: Optional[int] = None
    max_teachers: Optional[int] = None
    max_courses: Optional[int] = None
    is_active: Optional[bool] = None

class SubscriptionPlanRead(SubscriptionPlanBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

class SubscriptionBase(BaseModel):
    admin_id: int
    plan_id: int
    status: str  # active, cancelled, expired
    start_date: date
    end_date: date
    auto_renew: bool = True

class SubscriptionCreate(SubscriptionBase):
    pass

class SubscriptionUpdate(BaseModel):
    plan_id: Optional[int] = None
    status: Optional[str] = None
    end_date: Optional[date] = None
    auto_renew: Optional[bool] = None

class SubscriptionRead(SubscriptionBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

class SubscriptionInvoiceBase(BaseModel):
    subscription_id: int
    amount: float
    status: str  # paid, pending, failed
    billing_date: date
    paid_date: Optional[date] = None
    payment_method: Optional[str] = None
    invoice_number: str

class SubscriptionInvoiceCreate(SubscriptionInvoiceBase):
    pass

class SubscriptionInvoiceUpdate(BaseModel):
    status: Optional[str] = None
    paid_date: Optional[date] = None
    payment_method: Optional[str] = None

class SubscriptionInvoiceRead(SubscriptionInvoiceBase):
    id: int
    created_at: datetime

    model_config = {"from_attributes": True}

class UsageMetricsBase(BaseModel):
    admin_id: int
    metric_name: str
    metric_value: int

class UsageMetricsCreate(UsageMetricsBase):
    pass

class UsageMetricsRead(UsageMetricsBase):
    id: int
    recorded_at: datetime

    model_config = {"from_attributes": True}